### DISCUSSION TITLE
How to define the grain boundary area between grains

### DISCUSSION THREAD

**Discussion by Nuwanted** [idaholab/moose#22870] [Q&A Modules: Phase field] [answered]
Dear MOOSE experts,
     Recently, I wanted to simulate the evolution of the void in the polycrystalline material, and different nonconservative order parameters are used to represent different grains. When I use the block function to distinguish different grains and give them different surface mobility, void cannot evolve. Why? Is it because the grain boundary area is not defined?
     Any suggestions or recommendations to fix the problem would be greatly appreciated.

**Comment by GiudGiud** [comment]
@permcody

**Reply by friedmud** [reply]
Bumping @permcody again.

**Comment by permcody** [comment, marked as answer]
I'm not sure I understand exactly what you are doing here. Are you starting from some of our polycrystal examples or tests? We have a robust capability that allows you to identify each grain separately so that you can apply different attributes to each grain, which affect the Cahn-Hillard equations, which in turn impact your grain evolution. Normally in phase-field, we don't directly define the grain boundary area. Those quantities are calculated as by-product of solving the phase-field equations. We have a `VectorPostprocessor` that can inspect your converged solution and can calculate those quantities. I recommend you take a look at the examples here: https://github.com/idaholab/moose/tree/next/modules/phase_field/examples
and perhaps the integration tests as well: https://github.com/idaholab/moose/tree/next/modules/phase_field/test/tests

**Reply by Nuwanted** [reply]
Thank you for your reply, I mean that I use two nonconservative variables eta1 and eta2 to represent two different grains. The internal variable value of the grain is 1, the external variable value is 0, and the grain boundary changes from 1 to 0. I want to distinguish different grains by variable values. Some of my code is as follows:

InputParameters
ANisotropicForth::validParams()
{
	Params.addRequiredCoupledVar("nop1", "nonconservtion Order parameter 1 ");
	Params.addRequiredCoupledVar("nop2", "nonconservtion Order parameter 2 ");
	return Params;
	return Params;
}
ANisotropicForth::ANisotropicForth(const InputParameters & Parameters)
	: Material(Parameters),
	// Get the parameters from the input file
       _nop1(coupledValue("nop1")),
	_nop2(coupledValue("nop2"))
if (0 < _nop1 < 1 & _nop2 = 0 & _nop1 = 0)
	{
		// Calculate  function of mobility

		_fun[_qp] = ((_delta1 * std::pow(cos(_j1 * (angle + _theta0 * libMesh::pi / 180)), 2.0) + 1.0) / (1.0 + _delta1));
Nop1 and nop2 represent eta1 and eta2 respectively. When I compile, Moose prompts me that nop does not match the<symbol. Why?

**Reply by GiudGiud** [reply]
I've never written 0 < _nop1 < 1 in C++. Is that allowed?
This is definitely problematic:
if (..... _nop2 = 0)
it should be ==

But it does not make sense to check if nop1 is between 0 and 1, excluding 0, and at the same time nop1 =0 . These are mutually exclusive

**Reply by permcody** [reply]
`0 < _nop1 < 1` is an expression, but definitely not the one you want for any language. Your compiler should be screaming at you. Please fix your logic and see if that fixes your issue. You need to split that expression into two expressions. Also "&" should be "&&". I count no fewer than 6 bugs in that one line of code!

**Reply by Nuwanted** [reply]
Sorry, I am not familiar with the writing of C language, I will carefully modify my code, thank you very much for your reply.
